src/adapter/src/catalog.rs

```diff
@@ -5192,6 +5235,65 @@ impl Catalog {
                 Op::DropTimeline(timeline) => {
                     tx.remove_timestamp(timeline);
                 }
+                Op::GrantRole {
+                    role_id,
+                    member_id,
+                    grantor_id,
+                } => {
+                    let existing_role = state.get_role_mut(&member_id);
+                    if is_reserved_name(&existing_role.name) {
+                        return Err(AdapterError::Catalog(Error::new(
+                            ErrorKind::ReservedRoleName(existing_role.name.clone()),
+                        )));
+                    }
+                    existing_role.membership.map.insert(role_id, grantor_id);
```

Comment:
IMO we should also check for circular membership here. I see that we check for it when planning, but since we consider circular membership to be invalid then it would be nice to also make it an invalid state in the data structure that represents it

Reply:
Good point, I just moved the check to here.